### PR TITLE
fix(oidc-client): fix bugs and add more tests with refactor

### PR DESCRIPTION
**Pull request by IzumiSy**
https://tailor.atlassian.net/jira/software/c/projects/FL/boards/22?selectedIssue=FL-115

# Background

Integrating oidc-client in IMA made us encounter some critical bugs 

# Changes

- Fix a bug that `useSession` does not redirect users to the unauthorized route if guarded.
- Add `middlewareRouter` that is responsible for routing intercepted requests in Next.js middleware.
- Add more tests.
- Fix typo in README.

### CODE REVIEW

File: packages/oidc-client/src/server/middleware.test.ts [hunk 1]
```diff
@@ -1,29 +1,21 @@
...
-
-  beforeEach(() => {
-    fetchSpy = vi
-      .spyOn(global, "fetch")
-      .mockResolvedValue(new Response(JSON.stringify(mockSession)));
-  });
-
-  afterEach(() => {
-    fetchSpy.mockRestore();
-  });
```
Comment (IzumiSy) [inline packages/oidc-client/src/server/middleware.test.ts:25]: Mocking `fetch` is not really good so use msw here as well

File: packages/oidc-client/package.json [hunk 2]
```diff
@@ -41,6 +41,7 @@
     "@types/react-dom": "^18.2.17",
     "@typescript-eslint/eslint-plugin": "^6.13.2",
     "@typescript-eslint/parser": "^6.13.2",
+    "@vitejs/plugin-react": "^4.2.1",
```
Comment (IzumiSy) [inline packages/oidc-client/package.json:44]: Sometimes tests in oidc-client fails with `React is not defiend` error, so installed this to suppress the error.

File: packages/oidc-client/README.md [hunk 3]
```diff
@@ -144,7 +144,7 @@ const Component = async () => {
   const { token } = useSession();
   const { getCurrentUser } = usePlatform();
 
-  const user = await getUser(token);
+  const user = await getCurrentUser(token);
```
Comment (IzumiSy) [inline packages/oidc-client/README.md:147]: typo

File: packages/oidc-client/src/client/hooks.ts [hunk 4]
```diff
@@ -71,23 +80,30 @@ export const usePlatform = () => {
...
+    internalClientSession = session;
   };
 
-  if (!useSessionResult) {
+  if (!internalClientSession) {
     throw getSession();
   }
 
-  if (options?.required && useSessionResult.token === undefined) {
-    redirect(internalUnauthorizedPath);
```
Comment (IzumiSy) [inline packages/oidc-client/src/client/hooks.ts:89]: `redirect` is not available in client components, so here is replaced with `window.location.replace` in new changes.

Comment (IzumiSy) [reply packages/oidc-client/src/client/hooks.ts:89]: Plus, this guard checking should be done before suspending, so moved this to the beginning of this hook.

File: packages/oidc-client/src/lib/config.ts [hunk 5]
```diff
@@ -48,13 +48,3 @@ export class Config {
...
-
-// Internal path to fetch token from client components
-// `useSession` hook fetches token from this endpoint by extracting it from cookies.
-export const clientSessionPath = "/__auth/session" as const;
-
-// Internal path redirecting (through middleware) to `unauthorizedPath` set in ContextConfig
-// This is the path used in redirection caused by login guard in server components
-// The middleware that reads ContextConfig is in charge of redirecting to the `unauthorizedPath`
-// through this path when unauthorized users guarded in `getServerSession`.
-export const internalUnauthorizedPath = "/__auth/unauthorized" as const;
```
Comment (IzumiSy) [inline packages/oidc-client/src/lib/config.ts:60]: Moved these to `server/middleware/internal`.

File: packages/oidc-client/src/server/middleware.ts [hunk 6]
```diff
@@ -1,108 +1,85 @@
...
+export type RouteHandler = (
+  // Removing `options` field here is because it accepts `onError` callback,
+  // but calling onError should only be middlewareRouter itself for simplicity,
+  // not in each middleware handler.
+  params: Omit<RouterParams, "options"> & {
+    options?: MiddlewareHandlerOptions;
+  },
+) => Promise<NextResponse<unknown> | undefined> | NextResponse<unknown>;
 
-export const handleCallback = async (
```
Comment (IzumiSy) [inline packages/oidc-client/src/server/middleware.ts:57]: This handler for callback route is moved to `server/middleware/callback`

File: packages/oidc-client/src/tests/mocks.ts [hunk 7]
```diff
@@ -7,12 +10,51 @@ export const mockSession: Session = {
   user_id: "43a05b99-ebe1-4b89-8284-e4447e3a3551",
 };
 
-export const mockAuthConfig = new Config({
+const defaultMockedResponse = {
```
Comment (jfeltesse) [inline packages/oidc-client/src/tests/mocks.ts:13]: for consistency how about naming this `defaultMockResponse` (without "ed")

File: packages/oidc-client/package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@tailor-platform/oidc-client",
-  "version": "0.5.0",
+  "version": "0.6.0-preview.0",
```
Comment (jfeltesse) [inline packages/oidc-client/package.json:3]: need to remove `-preview.0`

File: packages/oidc-client/src/tests/mocks.ts (hunk 7 above)
Comment (IzumiSy) [reply packages/oidc-client/src/tests/mocks.ts:13]: Agreed. Updated!

File: packages/oidc-client/package.json (hunk 8 above)
Comment (IzumiSy) [reply packages/oidc-client/package.json]: Removed 🙏

**Review by jfeltesse — APPROVED** [review]

**Pull request merged**